### ISSUE TITLE
ACIA Prop->Z80 becomes borked (buffer issue) when keys are pressed while Z80 has not been reset

### ISSUE THREAD

**Issue by Wuerfel21** [feilipu/ux_module#7]
Seems to buffer each character (such that it comes out when the second next character is put in). I've isolated it to the ACIA Tx stuff, IDK what's up with that.

**Comment by Wuerfel21** [comment]
This effect isn't present in my "factory firmware", but does show up there, too, if I press ctrl+alt+del (albeit with a different (much longer) buffer depth)

strange...

If you remember, could you please tell me which comment my "factory firmware" was built from?

**Comment by Wuerfel21** [comment]
No wait, I'm just stupid!

It happens to the same extent in both versions and is actually triggered by pressing any keys before resetting the Z80 (on my setup, the Z80 apparently has to be reset every time the UX module boots, even if it was already running, for anything to happen. Similiar story with the ACIA card though, so I guess that's normal)

**Comment by Wuerfel21** [comment]
I think I figured the issue out: The receive interrupt is triggered when a byte is entered into the tx_buffer and the Z80 code only reads one byte per interrupt. So any bytes entered while the interrupt isn't being serviced just clog up the buffer.

**Comment by Wuerfel21** [comment]
Altough with the way the ACIA emulation has to wait on the bus, I don't really know where else to trigger the interrupt

**Comment by feilipu** [comment]
The ACIA emulation has to trigger a new interrupt, if there are [still bytes remaining in the Tx buffer](https://github.com/feilipu/ux_module/blob/main/src/acia_rc2014.spin#L385).

But, it isn't working totally correctly now, because if two bytes come into the Prop Tx buffer (ACIA emulation) together, then only one interrupt is serviced, and I guess that's because the state machine is very crappy. Need to work on this too.

The ACIA Module interrupt servicing code on the Z80 side could be fixed to always look for more bytes (by reading the ACIA status byte again after the byte has been read), but this would assume that the code expects something to be there and for the real ACIA there can't be anything there. In real hardware ACIA the received byte would be overwritten (and there'd be overflow set in the status), or the hardware flow control would be asked to manage this (more likely).

This is the "standard" code by Searle for handling the byte reception. It doesn't check that there are no more bytes available after it has handled one byte. I do [something similar](https://github.com/RC2014Z80/RC2014/blob/master/ROMs/HexLoadr/int32k.asm#L54) too.

I could fix the issue for myself simply by checking the status register again before moving on to handle Tx, but then it would break the requirement to be drop in compatible. So the Prop Tx side will have to be fixed to generate 1 interrupt for 1 byte, somehow.

```
serialInt:      PUSH     AF
                PUSH     HL

                IN       A,($80)
                AND      $01             ; Check if interupt due to read buffer full
                JR       Z,rts0          ; if not, ignore

                IN       A,($81)
                PUSH     AF
                LD       A,(serBufUsed)
                CP       SER_BUFSIZE     ; If full then ignore
                JR       NZ,notFull
                POP      AF
                JR       rts0

notFull:        LD       HL,(serInPtr)
                INC      HL
                LD       A,L             ; Only need to check low byte becasuse buffer<256 bytes
                CP       (serBuf+SER_BUFSIZE) & $FF
                JR       NZ, notWrap
                LD       HL,serBuf
notWrap:        LD       (serInPtr),HL
                POP      AF
                LD       (HL),A
                LD       A,(serBufUsed)
                INC      A
                LD       (serBufUsed),A
                CP       SER_FULLSIZE
                JR       C,rts0
                LD       A,RTS_HIGH
                OUT      ($80),A
rts0:           POP      HL
                POP      AF
                EI
                RETI
```

So to do the emulation properly, I'll need to think more about this issue... I think perhaps the fresh interrupt generated just needs a bit more delay, so the Z80 can recognise it?

**Comment by feilipu** [comment]
I think this issue is partially fixed in my implementations, but I think it can't be fully fixed using the default ROM.

The issue is that the real ACIA device is double buffered, and the Z80 side needs to check that no new byte has been received during the interrupt processing, and if a new byte has been received it needs to go back to get it. The default MS Basic ROM doesn't do this, nor does it read the overrun flag (which would be another way to fix this).

On the Propeller side it may be possible to fudge an answer by delaying transmission of bytes for a fixed time, but that would slow things down substantially, and still not be a definitive fix.

I've revised [my MS Basic](https://github.com/feilipu/NASCOM_BASIC_4.7/commit/4ea30811e382696399ed06704ddbbe22f8134a9e) and [CP/M-IDE](https://github.com/RC2014Z80/RC2014/commit/6ff29009befe3cb31755104b886b5711c0d345e6) implementations to fix the overrun and it was already properly implemented in [RomWBW](https://github.com/wwarthen/RomWBW/blob/master/Source/HBIOS/acia.asm#L242).

**Comment by Wuerfel21** [comment]
I guess a workaround would be to generate another interrupt if the last one hasn't been acknowledged for a while (250ms or so)

**Comment by feilipu** [comment]
I've added a `txFlush` that should resolve this issue of characters clogging the Tx queue before the Z80 is booted.

The separate issue of the Prop being too fast for the Z80 with characters has also been handled by managing the ACIA double buffer, but imperfectly (not) for standard firmware. The spin code toggling of the Z80 Rx interrupt is quite slow (fortunately) so there should be heaps of time for the Z80 to handle characters arriving in normal circumstances.
